src/operator/numpy/np_indexing_op.cu

```diff
@@ -0,0 +1,574 @@
+                                    prefix_sum,
+                                    idx_size,
+                                    stream);
+    size_t buffer_size = idx_size * sizeof(int32_t);
+    temp_storage_bytes += buffer_size;
+    // Allocate memory on GPU and allocate pointer
+    Tensor<gpu, 1, char> workspace =
+        ctx.requested[0].get_space_typed<gpu, 1, char>(Shape1(temp_storage_bytes), s);
+    prefix_sum = reinterpret_cast<int32_t*>(workspace.dptr_);
+    d_temp_storage = workspace.dptr_ + buffer_size;
+    MSHADOW_TYPE_SWITCH_WITH_BOOL(idx.dtype(), IType, {
+        mxnet_op::Kernel<mshadow_op::identity_with_cast, gpu>::Launch(
+        s, idx.shape()[0], prefix_sum, idx.data().dptr<IType>());
+    });
+    // Calculate prefix sum
+    cub::DeviceScan::InclusiveSum(d_temp_storage,
+                                    temp_storage_bytes,
+                                    prefix_sum,
+                                    prefix_sum,
+                                    idx_size,
+                                    stream);
+    CUDA_CALL(cudaMemcpyAsync(&valid_num, &prefix_sum[idx_size - 1], sizeof(int32_t),
+                                cudaMemcpyDeviceToHost, stream));
+    CUDA_CALL(cudaStreamSynchronize(stream));
+
+    // Set the output shape forcefully
+    mxnet::TShape data_shape = data.shape();
+    data_shape[axis] = valid_num;
+    const_cast<NDArray &>(out).Init(data_shape);
+    size_t input_size = data.shape().Size();
+    size_t col_size = input_size / idx.shape()[0];
+    // Do the copy
+    MSHADOW_TYPE_SWITCH_WITH_BOOL(out.dtype(), DType, {
+        if (valid_num > 0) {
+        mxnet_op::Kernel<BooleanMaskForwardKernel, gpu>::Launch(
+            s, input_size, out.data().dptr<DType>(),
+            data.data().dptr<DType>(), prefix_sum, col_size);
+        }
+    });
+} else if (inputs[np_indexing_::kIdx].dtype() == mshadow::kInt8 ||
+           inputs[np_indexing_::kIdx].dtype() == mshadow::kInt16 ||
+           inputs[np_indexing_::kIdx].dtype() == mshadow::kInt32 ||
+           inputs[np_indexing_::kIdx].dtype() == mshadow::kInt64) {
+    using namespace mxnet_op;
+    const mxnet::TShape& idxshape = inputs[np_indexing_::kIdx].shape();
+    const mxnet::TShape& arrshape = inputs[np_indexing_::kArr].shape();
+
+    if (idxshape.Size() == 0) {
+        return;
+    }
+
+    mxnet::TShape oshape(idxshape.ndim() + arrshape.ndim() - 1, -1);
+    for (index_t i = 0; i < idxshape.ndim(); ++i) {
+        oshape[i] = idxshape[i];
+    }
+    for (index_t i = 0; i < arrshape.ndim(); i++) {
+        if (i < 0) {
+        oshape[i] = arrshape[i];
+        } else if (i > 0) {
+        oshape[i + idxshape.ndim() - 1] = arrshape[i];
+        }
+    }
+
+    const NDArray &out = outputs[0];
+    const_cast<NDArray &>(out).Init(oshape);
+
+    Stream<gpu> *s = ctx.get_stream<gpu>();
+
+    MSHADOW_TYPE_SWITCH_WITH_BOOL(outputs[np_indexing_::kOut].dtype(), DType, {  // output data type
```

Comment:
I think we may use the `MSHADOW_TYPE_SWITCH` version without bool. Because the bool type won't be used.